// angular2-ladda/esm5/lib/ladda.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Optional, Inject } from '@angular/core';
import { LaddaConfig, LaddaConfigArgs, configAttributes } from './ladda-config';
import { create as createLadda } from 'ladda';
var LaddaDirective = /** @class */ (function () {
    function LaddaDirective(el, config) {
        this.el = el.nativeElement;
        if (!config) {
            return;
        }
        // apply default styles if they aren't overwritten by an attribute
        for (var attribute in configAttributes) {
            /** @type {?} */
            var configValue = config[configAttributes[attribute]];
            if (!configValue) {
                continue; // don't waste time reading the attribute
            }
            if (!this.el.getAttribute(attribute)) {
                // attribute isn't set - apply the default config value
                /** @type {?} */
                var value = (typeof configValue === "number") ? configValue.toString() : configValue;
                this.el.setAttribute(attribute, value);
            }
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    LaddaDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this.ladda) {
            return; // needed since ngOnChanges is called before ngOnInit
        }
        if (changes.loading) {
            this.updateLadda(changes.loading.previousValue);
        }
        if (changes.disabled) {
            this.updateDisabled();
        }
    };
    /**
     * @return {?}
     */
    LaddaDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.ladda = createLadda(this.el);
        // if the initial loading value isn't false, a timeout of 0 ms
        // is necessary for the calculated spinner size to be correct.
        setTimeout((/**
         * @return {?}
         */
        function () { _this.updateLadda(false); }), 0);
    };
    /**
     * @return {?}
     */
    LaddaDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.ladda) {
            this.ladda.remove();
        }
    };
    /**
     * @private
     * @param {?} previousValue
     * @return {?}
     */
    LaddaDirective.prototype.updateLadda = /**
     * @private
     * @param {?} previousValue
     * @return {?}
     */
    function (previousValue) {
        /** @type {?} */
        var loading = typeof this.loading === 'number' || !!this.loading;
        /** @type {?} */
        var wasLoading = typeof previousValue === 'number' || !!previousValue;
        if (!loading) {
            if (wasLoading) {
                this.ladda.stop();
            }
            return this.updateDisabled();
        }
        if (!wasLoading) {
            this.ladda.start();
        }
        if (typeof this.loading === 'number') {
            this.ladda.setProgress(this.loading);
        }
    };
    /**
     * @private
     * @return {?}
     */
    LaddaDirective.prototype.updateDisabled = /**
     * @private
     * @return {?}
     */
    function () {
        this.el.disabled = this.disabled;
    };
    LaddaDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ladda]'
                },] }
    ];
    /** @nocollapse */
    LaddaDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: LaddaConfigArgs, decorators: [{ type: Inject, args: [LaddaConfig,] }, { type: Optional }] }
    ]; };
    LaddaDirective.propDecorators = {
        loading: [{ type: Input, args: ['ladda',] }],
        disabled: [{ type: Input }]
    };
    return LaddaDirective;
}());
export { LaddaDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LaddaDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    LaddaDirective.prototype.ladda;
    /** @type {?} */
    LaddaDirective.prototype.loading;
    /** @type {?} */
    LaddaDirective.prototype.disabled;
}
